### ISSUE TITLE
Rack::Timeout#call running for 12+ minutes on Heroku

### ISSUE THREAD

**Issue by monfresh** [heroku/rack-timeout#73]
Hi,

I've been seeing some strange timeouts on Heroku ever since I upgraded to the latest rack-timeout.

Here is a screenshot from New Relic that shows `Rack::Timeout#call` being responsible for the high response time:
![new relic](http://cl.ly/image/460x2N3q3k3l/download/Screen%20Shot%202015-04-03%20at%2012.45.53%20AM.png)

And here is the app: https://github.com/smcgov/ohana-api-smc

I'm deleting the gem for now.

**Comment by kch** [comment]
upgraded from what version to what version?

**Comment by kch** [comment]
Rack-timeout is unable to abort things hung in C land, like a bunch of IO stuff, so that would be my first suspicion.

Those requests should eventually time out, raise, and the backtrace might show what was holding the request.

**Comment by monfresh** [comment]
I upgraded from 0.2.0 to 0.2.1, and then from 0.2.1 to 0.2.4. Most of my [commits over the past few days](https://github.com/smcgov/ohana-api-smc/commits/master) have been upgrading gems, and I also upgraded from Rails 4.1.9 to 4.2.1, and upgraded Ruby from 2.1.5 to 2.2.1. I haven't made any relevant changes to the API code, which is why I thought it was `rack-timeout` that was causing this. This is the first time I've ever seen such long requests in my app.

Logentries reported four "High Response Time" incidents yesterday where the request kept lingering for about 4.5 hours. Since I removed `rack-timeout` last night, I haven't heard a peep from Logentries, but that's not necessarily the cause. I'll let you know how it goes over the next few days.

Also, Logentries does not show any stack traces around these incidents. Here is a sample log entry:

```
» 3 Apr 2015 00:33:08.641 2015-04-03 04:33:08.487028+00:00 heroku web.1 - - Stopping all processes with SIGTERM
» 3 Apr 2015 00:33:09.692 2015-04-03 04:33:09.384968+00:00 app web.1 - - source=rack-timeout id=428a23bb-52b2-419d-a780-8ae7cb31de7f wait=14ms timeout=10000ms state=ready
» 3 Apr 2015 00:33:09.702 2015-04-03 04:33:09.389091+00:00 app web.1 - - Started GET "/" for 50.31.164.139/NX at 2015-04-03 04:33:09 +0000
» 3 Apr 2015 00:33:09.782 2015-04-03 04:33:09.392177+00:00 app web.1 - - ** [NewRelic][04/03/15 04:33:09 +0000 web.1 (14)] INFO : Starting Agent shutdown
» 3 Apr 2015 00:33:09.782 2015-04-03 04:33:09.451291+00:00 app web.1 - - Completed 200 OK in 1ms (Views: 0.4ms | ActiveRecord: 0.0ms)
» 3 Apr 2015 00:33:09.782 2015-04-03 04:33:09.503370+00:00 app web.1 - - cache: [HEAD /] miss
» 3 Apr 2015 00:33:09.782 2015-04-03 04:33:09.509550+00:00 app web.1 - - source=rack-timeout id=61afdafc-198f-484e-a93f-682b2c4685ca wait=8ms timeout=10000ms state=ready
» 3 Apr 2015 00:33:09.782 2015-04-03 04:33:09.582708+00:00 app web.1 - - Processing by Api::V1::LocationsController#show as JSON
» 3 Apr 2015 00:33:09.790 2015-04-03 04:33:09.434443+00:00 app web.1 - - Processing by Api::V1::RootController#index as JSON
» 3 Apr 2015 00:33:09.790 2015-04-03 04:33:09.493012+00:00 app web.1 - - [3] - Gracefully shutting down workers...

3 Apr 2015 00:33:09.790 2015-04-03 04:33:09.504070+00:00 app web.1 - - source=rack-timeout id=428a23bb-52b2-419d-a780-8ae7cb31de7f wait=14ms timeout=10000ms service=15905014ms state=completed High Response Time

» 3 Apr 2015 00:33:09.790 2015-04-03 04:33:09.535648+00:00 app web.1 - - Started GET "/locations/home-health-and-hospice-programs-san-francisco?api_token=[FILTERED]" for 54.90.138.78/ec2-54-90-138-78.compute-1.amazonaws.com at 2015-04-03 04:33:09 +0000
» 3 Apr 2015 00:33:09.790 2015-04-03 04:33:09.609298+00:00 app web.1 - - Parameters: {"api_token"=>"[FILTERED]", "id"=>"home-health-and-hospice-programs-san-francisco"}
» 3 Apr 2015 00:33:09.974 2015-04-03 04:33:09.679731+00:00 app web.1 - - Location Load (30.3ms) SELECT "locations".* FROM "locations" WHERE "locations"."slug" = $1 ORDER BY "locations"."id" ASC LIMIT 1 [["slug", "home-health-and-hospice-programs-san-francisco"]]
» 3 Apr 2015 00:33:10.057 2015-04-03 04:33:09.761968+00:00 app web.1 - - Contact Load (77.6ms) SELECT "contacts".* FROM "contacts" WHERE "contacts"."location_id" IN (1242) ORDER BY id ASC
» 3 Apr 2015 00:33:10.063 2015-04-03 04:33:09.813236+00:00 app web.1 - - Phone Load (46.0ms) SELECT "phones".* FROM "phones" WHERE "phones"."contact_id" IN (1442, 1443) ORDER BY id ASC
» 3 Apr 2015 00:33:10.063 2015-04-03 04:33:09.929259+00:00 app web.1 - - HABTM_Categories Load (61.3ms) SELECT "categories_services".* FROM "categories_services" WHERE "categories_services"."service_id" IN (1242)
» 3 Apr 2015 00:33:10.145 2015-04-03 04:33:09.853206+00:00 app web.1 - - Service Load (30.1ms) SELECT "services".* FROM "services" WHERE "services"."location_id" IN (1242)
» 3 Apr 2015 00:33:10.442 2015-04-03 04:33:10.147040+00:00 app web.1 - - Category Load (27.8ms) SELECT "categories".* FROM "categories" WHERE "categories"."id" IN (78, 87, 88, 90, 91, 101, 104, 107, 109) ORDER BY taxonomy_id asc
» 3 Apr 2015 00:33:10.509 2015-04-03 04:33:10.214822+00:00 app web.1 - - Contact Load (22.5ms) SELECT "contacts".* FROM "contacts" WHERE "contacts"."service_id" IN (1242) ORDER BY id ASC
» 3 Apr 2015 00:33:10.531 2015-04-03 04:33:10.266270+00:00 app web.1 - - Phone Load (32.5ms) SELECT "phones".* FROM "phones" WHERE "phones"."service_id" IN (1242) ORDER BY id ASC
» 3 Apr 2015 00:33:10.531 2015-04-03 04:33:10.381297+00:00 app web.1 - - HolidaySchedule Load (34.6ms) SELECT "holiday_schedules".* FROM "holiday_schedules" WHERE "holiday_schedules"."service_id" IN (1242)
» 3 Apr 2015 00:33:10.597 2015-04-03 04:33:10.314982+00:00 app web.1 - - RegularSchedule Load (46.9ms) SELECT "regular_schedules".* FROM "regular_schedules" WHERE "regular_schedules"."service_id" IN (1242) ORDER BY weekday ASC
» 3 Apr 2015 00:33:10.777 2015-04-03 04:33:10.482437+00:00 app web.1 - - Completed 304 Not Modified in 786ms (ActiveRecord: 409.5ms)

» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.547163+00:00 app web.1 - - source=rack-timeout id=61afdafc-198f-484e-a93f-682b2c4685ca wait=8ms timeout=10000ms service=15858400ms state=completed High Response Time

» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.571773+00:00 app web.1 - - Started GET "/" for 54.247.188.179/ec2-54-247-188-179.eu-west-1.compute.amazonaws.com at 2015-04-03 04:33:10 +0000
» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.590343+00:00 app web.1 - - Completed 200 OK in 1ms (Views: 0.3ms | ActiveRecord: 0.0ms)

» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.625052+00:00 app web.1 - - source=rack-timeout id=7c119c2a-062f-428a-8d2b-f81356b4efe2 wait=4ms timeout=10000ms service=15688305ms state=completed High Response Time

» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.649805+00:00 app web.1 - - Started GET "/" for 54.248.250.232/ec2-54-248-250-232.ap-northeast-1.compute.amazonaws.com at 2015-04-03 04:33:10 +0000
» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.656688+00:00 app web.1 - - Completed 200 OK in 4ms (Views: 0.4ms | ActiveRecord: 0.0ms)

» 3 Apr 2015 00:33:10.846 2015-04-03 04:33:10.658994+00:00 app web.1 - - source=rack-timeout id=672cd6fe-32c3-4426-a2cf-f76463f2da1a wait=6ms timeout=10000ms service=15811184ms state=completed High Response Time

» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.534825+00:00 app web.1 - - cache: [GET /locations/home-health-and-hospice-programs-san-francisco?api_token=[FILTERED]] stale, valid, store
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.569571+00:00 app web.1 - - source=rack-timeout id=7c119c2a-062f-428a-8d2b-f81356b4efe2 wait=4ms timeout=10000ms state=ready
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.582872+00:00 app web.1 - - Processing by Api::V1::RootController#index as JSON
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.602401+00:00 app web.1 - - cache: [HEAD /] miss
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.647219+00:00 app web.1 - - source=rack-timeout id=672cd6fe-32c3-4426-a2cf-f76463f2da1a wait=6ms timeout=10000ms state=ready
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.652788+00:00 app web.1 - - Processing by Api::V1::RootController#index as JSON
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.658559+00:00 app web.1 - - cache: [HEAD /] miss
» 3 Apr 2015 00:33:10.866 2015-04-03 04:33:10.660875+00:00 app web.1 - - ** [NewRelic][04/03/15 04:33:10 +0000 web.1 (6)] INFO : Starting Agent shutdown
» 3 Apr 2015 00:33:13.920 2015-04-03 04:33:13.533282+00:00 app web.1 - - [3] === puma shutdown: 2015-04-03 04:33:13 +0000 ===
» 3 Apr 2015 00:33:14.008 2015-04-03 04:33:13.533289+00:00 app web.1 - - [3] - Goodbye!
» 3 Apr 2015 00:33:16.104 2015-04-03 04:33:15.808817+00:00 heroku web.1 - - Process exited with status 0
```

**Comment by Xanthus** [comment]
I get this too. I've started working on an app targeting Heroku and rack-timeout is causing tests and the dev server to randomly freeze during requests. Haven't run any traces or anything yet. Using rack-timeout 0.2.1, Rails 4.2.1 and Ruby 2.2.1. I'll try 0.2.4 too.

**Comment by kch** [comment]
Puma?

**Comment by monfresh** [comment]
Just wanted to follow up and mention that I was (and still am) using Puma when I reported this. Also, this is just an observation, and there could be many reasons for these results, but when I had rack-timeout, I used to get frequent "high response time" alerts from logentries (mostly between 5s and 10s). After I removed rack-timeout, they disappeared. My logs only go back 7 days, and in the past week I haven't had a single high response alert.

**Comment by kch** [comment]
Yeah it seems rack-timeout and puma don't play nice in many ways. Gotta take some time to investigate that.

**Comment by Xanthus** [comment]
Yip, also on Puma here.

**Comment by jhjwind** [comment]
+1

**Comment by apeacox** [comment]
+1 on Puma

**Comment by sharq1** [comment]
+1 on Puma

**Comment by mensfeld** [comment]
+1 on Puma

**Comment by kch** [comment]
This is rather likely due to #76. Watch #78 for a beta with the fix from #86.
